### PR TITLE
Separate string solving code

### PR DESCRIPTION
**Pull request by Adda0**
This PR moves string solving code outside `nfa.hh` and to its own namespace `StringSolving`. The string solving code should be kept hidden from the public interface of libMata, unless the user explicitly asks for it by including `nfa-string-solving.hh` and accessing the code from within its namespace.

When we figure out whether to keep the string solving code in Mata, more changes can be introduced. The main goal here is to stabilize the interface of `nfa.hh`.

The moved string solving operations comprise:
- segmentation,
- computation of shortest words, and
- noodlification.

### CODE REVIEW

File: src/nfa/string-solving/nfa-string-solving.cc [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+using namespace Mata::Nfa;
+using namespace Mata::Nfa::StringSolving;
+
+namespace {
+/**
+ * Get a number of permutations for computed epsilon depths.
+ * @param[in] epsilon_depths Computed list of epsilon transitions for each depth.
+ * @return Number of permutations.
+ */
+    size_t get_num_of_permutations(const SegNfa::Segmentation::EpsilonDepthTransitions& epsilon_depths)
```
Comment (martinhruska) [inline src/nfa/string-solving/nfa-string-solving.cc:29]: please fix indent.

File: include/mata/nfa.hh [hunk 2]
```diff
@@ -732,7 +724,7 @@ public:
      * @return Returns reference element of transition list with epsilon transitions or end of transition list when
      * there are no epsilon transitions.
      */
-    Moves::const_iterator get_epsilon_transitions(State state, const Symbol epsilon = EPSILON) const;
+    Moves::const_iterator get_epsilon_transitions(State state, Symbol epsilon = EPSILON) const;
```
Comment (martinhruska) [inline include/mata/nfa.hh:727]: why is parameter no longer constant?

**Review by martinhruska — APPROVED** [review]

**Comment by martinhruska** [comment]
I have no problem with PR. It just moves code related to string solving to special folder and does not break anything.

**Comment by martinhruska** [comment]
I would only consider one thing. Should the namespace `StringSolving` be subspace of `Nfa`. I think it makes sense to make it directly subspace of the `Mata` namespace.

File: src/nfa/string-solving/nfa-string-solving.cc (hunk 1 above)
Comment (Adda0) [reply src/nfa/string-solving/nfa-string-solving.cc:29]: Done.

File: include/mata/nfa.hh (hunk 2 above)
Comment (Adda0) [reply include/mata/nfa.hh:727]: It is constant (see the function definition), but `const` in function declaration has no functionality. The compiler omits it.

Comment (martinhruska) [reply include/mata/nfa.hh:727]: ok, legit

**Comment by Adda0** [comment]
> > I would only consider one thing. Should the namespace StringSolving be subspace of Nfa. I think it makes sense to make it directly subspace of the Mata namespace.
> 
> That is a good idea. I will refactor the code.

Aha, I take it back. I thought it was a good idea at first, too. However, when I was doing the refactorization when creating this PR, I realized that the functions are directly extending the functionality of NFAs. For example, segmentation within `SegNfa` namespace is directly specifying a type of NFAs.

Therefore, it boils down to whether we want to have the following structure of namespaces:
 ```
- Mata
    - Nfa
        - StringSolving
            - SegNfa
            - ...
       - DifferentExtension
       - ...   
    - Afa
    - ... 
```
or rather:
 ```
- Mata
    - Nfa
       - DifferentExtension
       - ...   
    - Afa
    - ... 
    - StringSolving
        - Nfa
            - SegNfa
            - ...
        - ...
```

We should discuss this further. Maybe even in the meeting on Thursday.

**Comment by martinhruska** [comment]
Ok, if it is directly extend Nfa, it makes sense to make it subnamespace of Nfa.

**Comment by Adda0** [comment]
This is how it was implemented originally, too. I would keep the current implementation, at least for the time being.

**Review by tfiedor — APPROVED** [review]
Seems OK to me

**Comment by martinhruska** [comment]
Well, it looks that all  changes have passed so we can merge the PR.

**Comment by Adda0** [comment]
I quickly rechecked the introduced changes. From where I stand, the PR is done and ready for merge.

**Pull request merged**